File: 6.hs
module Main where

import Safe
import Data.Map (Map)
import Data.Maybe (maybe)
import Data.List
import Data.Tuple
import qualified Data.Map as M

type Tree = Map String String

split' :: Eq a => [a] -> a -> [a] -> ([a], [a])
split' ls c [] = (ls, [])
split' ls c (x:xs)
  | c == x = (ls, xs)
  | otherwise = split' (ls ++ [x]) c xs

split = split' []

edges :: String -> [(String, String)]
edges = map (split ')') . lines

parseTree :: String -> Tree
parseTree = M.fromList . map swap . edges

parentsOf :: Tree -> String -> [String]
parentsOf t s =
  case M.lookup s t of
    Nothing -> []
    Just p -> p : parentsOf t p

heightOf :: Tree -> String -> Int
heightOf t = length . parentsOf t

heightTo :: Tree -> String -> String -> Maybe Int
heightTo t root child
  | root == child = Just 0
  | otherwise = do
    p <- M.lookup child t
    (1 +) <$> heightTo t root p

leastUpperBound :: Tree -> String -> String -> Maybe String
leastUpperBound t l r =
  let lParents = parentsOf t l
      rParents = parentsOf t r
   in headMay (lParents `intersect` rParents)

distanceBetween :: Tree -> String -> String -> Maybe Int
distanceBetween t l r = do
  lub <- leastUpperBound t l r
  lh <- heightTo t lub l
  rh <- heightTo t lub r
  pure (lh + rh)

main :: IO ()
main = do
  t <- parseTree <$> readFile "6.txt"
  -- Part 1
  print . sum . map (heightOf t) $ M.keys t
  -- Part 2
  print . fmap (subtract 2) $ distanceBetween t "YOU" "SAN"

File: 6.rb
# frozen_string_literal: true

require 'set'

class Tree
  attr_reader :edges

  def initialize(edges)
    @edges = edges
    @parents = {}
  end

  def self.from_edges(edges)
    adj = {}
    edges.each do |parent, child|
      adj[child] = parent
    end

    new(adj)
  end

  def height(node)
    parents(node).length
  end

  def nodes
    edges.keys
  end

  def distance_between(l, r)
    lub = least_upper_bound(l, r)
    height_to(lub, l) + height_to(lub, r)
  end

  private

  def least_upper_bound(l, r)
    l_parents = parents(l)
    r_parents = parents(r)
    height = [l_parents.length, r_parents.length].max
    visited = Set.new

    height.times do |i|
      l = l_parents[i]
      r = r_parents[i]

      return l if visited.include?(l)

      visited << l if l

      return r if visited.include?(r)

      visited << r if r
    end
  end

  def height_to(parent, child)
    parents(child).index(parent)
  end

  def parents(node)
    return @parents[node] if @parents[node]

    parent = edges[node]
    @parents[node] = if parent
                       [node] + parents(parent)
                     else
                       []
                     end
  end
end

tree = Tree.from_edges(
  File.read('6.txt').split("\n").map { |line| line.split(')') }
)

# Part 1
puts tree.nodes.map { |p| tree.height(p) }.sum
# Part 2
puts tree.distance_between('YOU', 'SAN') - 2
